### ISSUE TITLE
Phaul - Live migration

### ISSUE THREAD

**Issue by sourav295** [checkpoint-restore/criu#437]
I am trying to migrate a docker container from one docker host to another. Is the criu-dev branch capable of carrying out docker live migration using phaul (the go version integrated in criu and not the python one )?

**Comment by avagin** [comment]
https://github.com/moby/moby/issues/12995

**Comment by sourav295** [comment]
Thanks a lot for your reply. I know about libcontainer; however, I don't want to use it because of certain networking concerns.

I am curious if I carry out live migration between two hosts using phaul. (Not the python version, since it's abandoned) Although, I like logic of that code.

**Comment by avagin** [comment]
Oops, I posted a wrong link. The right one is https://github.com/containerd/containerd/issues/1731

**Comment by ZikeWang** [comment]
I want to carry out container live migration using docker-ce, does p.haul now support docker-ce?

**Comment by xemul** [comment]
@ZikeWang , the question should be reversed :) Like this: "does docker support p.haul?", because p.haul is a library that should be used by container runtime to live-migrate containers. AFAIU docker has worse support for criu as it could, p.haul itself is not integrated.

However, @adrianreber is doing great work in integrating criu features into docker, definitely he has better understanding than me of what's happening there.

**Comment by ZikeWang** [comment]
@xemul thx a lot for your reply ! A little more question,  I‘m going to start research on container migration. if I wanna use p.haul, maybe I need to use container runtime such as containerd rather than docker-ce, or try writing the patch in Go for the integration with docker ???

**Comment by adrianreber** [comment]
@ZikeWang so @xemul is right. I am working on container migration.

runc, as you probably know, has support for container migration using pre-copy and post-copy migration. Recently I added pre-copy migration to LXD (https://fosdem.org/2018/schedule/event/containers_optimized_migration/ , https://lisas.de/~adrian/?p=1294) and currently I am looking at adding container migration to podman.

I am unsure what the correct steps to add optimized container migration to docker are. Looking at containerd is probably the right thing to do. My main problem is that I do not understand the relationship between all the projects above runc. That is why I am currently looking at podman which seems to be a clear runc based implementation.

Getting all the container migration optimizations from runc to containerd sounds like a good idea.

Although I think that p.haul in go is the right solution I actually did not use it in LXD. The existing infrastructure always makes it difficult to rewrite everything to use p.haul.
